src/frontend/src/scheduler/distributed/stage.rs

```diff
@@ -327,38 +341,62 @@ impl StageRunner {
         }
 
         // Merge different task streams into a single stream.
-        let all_streams = select_all(buffered_streams);
+        let mut all_streams = select_all(buffered_streams);
 
         // Process the stream until finished.
         let mut running_task_cnt = 0;
         let mut sent_signal_to_next = false;
-        #[for_await]
-        for status_res in all_streams {
-            // The status can be Running, Finished, Failed etc. This stream contains status from
-            // different tasks.
-            let status = status_res.map_err(|e| RpcError(e.into()))?;
-            use risingwave_pb::task_service::task_info::TaskStatus as TaskStatusProst;
-            if TaskStatusProst::from_i32(status.task_info.as_ref().unwrap().task_status)
-                == Some(TaskStatusProst::Running)
-            {
-                running_task_cnt += 1;
-                // The task running count should always less or equal than the registered tasks
-                // number.
-                assert!(running_task_cnt <= self.tasks.keys().len());
-                // All tasks in this stage have been scheduled. Notify query runner to schedule next
-                // stage.
-                if running_task_cnt == self.tasks.keys().len() {
-                    self.notify_schedule_next_stage().await?;
-                    sent_signal_to_next = true;
+        let mut shutdown_tx = shutdown_tx;
+        // This loop will stops once receive a stop message, otherwise keep processing status
+        // message.
+        loop {
+            tokio::select! {
+                    biased;
+                    _ = &mut shutdown_tx => {
+                    // Received shutdown signal from query runner, should send abort RPC to all CNs.
+                    // change state to aborted. Note that the task cancel can only happen after schedule all these tasks to CN.
+                    // This can be an optimization for future: How to stop before schedule tasks.
```

Comment:
The abort can only happen after we schedule all tasks to CN. I think it is enough, the time before schedule task is too short so I doubt that we can stop before create task.

Reply:
LGTM